frontend/src/utils/AuthContext.tsx

```diff
@@ -113,17 +117,70 @@ export const AuthProvider = ({ children }: AuthProviderProps) => {
     "/_error",
   ];
 
+  // Paths that can be accessed only by supervisors
+  const supervisorPaths = [
+    "/events/[eventid]/attendees",
+    "/events/[eventid]/edit",
+    "/events/create",
+  ];
+
+  // Paths that can be accessed only by admins
+  // TODO: /manage is not an actual path yet; replace this with whatever is the 
+  // eventual path for the manage website page
+  const adminPaths = ["/manage", "/users/view"];
+
   const router = useRouter();
   useEffect(() => {
     const path = router.asPath;
-    const unsubscribe = onAuthStateChanged(auth, (user) => {
-      if (!user && !publicPaths.includes(path)) {
-        router.replace("/login");
-        setIsAuthenticated(false);
-      } else if (user && authPaths.includes(path)) {
+    const unsubscribe = onAuthStateChanged(auth, async (user) => {
+      if (user && authPaths.includes(path)) {
         router.replace("/events/view");
+      } else if (user) {
+        // User is authenticated
+        // Check if user has custom claims
+        const idTokenResult = await user.getIdTokenResult();
+        const { claims } = idTokenResult;
+        console.log(claims);
+
+        if (claims) {
+          // Redirect based on user role
+          if (claims.admin) {
+            setRole("Admin");
+            // Admin can access any path
+            setIsAuthenticated(true);
+          } else if (claims.supervisor) {
+            setRole("Supervisor");
+            // Supervisors only denied paths exclusively for admins
+            if (!adminPaths.includes(path)) {
+              setIsAuthenticated(true);
+            } else {
+              router.replace("/events/view");
+            }
+          } else if (claims.volunteer) {
+            setRole("Volunteer");
+            // Volunteers cannot access adminPaths or supervisorPaths
+            if (!adminPaths.includes(path) && !supervisorPaths.includes(path)) {
+              setIsAuthenticated(true);
+            } else {
+              router.replace("/events/view");
+            }
+          } else {
+            router.replace("/login");
+            setIsAuthenticated(false);
+          }
+        } else {
```

Comment:
Not a huge fan of chained if statements, especially in a useEffect. See if you can tidy this logic up a little bit. It is almost non-ideal for authentication where things have to be deterministic. I recommended simplifying some of the conditions using `&&` or using case statements!

Reply:
I combined some of the conditions using <&&> to hopefully make it a bit more readable!